### PR TITLE
linux: support simple ignore patterns

### PR DESCRIPTION
**Pull request by paul-marechal**
This is a first pass at implementing some mechanism to filter paths.
Linux only.

Some pattern examples:

- `/absolute/path[/][*]`
- `./path/parts[/][*]`
- `**/path/parts[/][*]`
- `*path/parts[/][*]`
- `path/parts[/][*]`

The general idea is to do simple string matching and convert patterns
into path parts that can then be matched against a real path.

For example, `**/foo` gets converted to `/foo/` and we will try to find
this subtring in the following paths:

- `/a/b/foo/c/` matches.
- `/a/b/foobar/c/` does not match.

In order for the pattern to match `/a/b/foobar/c/` we need to write
`**/foo*` which will be converted to `/foo`. We now have:

- `/a/b/foo/c` matches.
- `/a/b/foobar/c` matches.

This is because we are not required to find `/` right after `/foo`.
Depending on if we use a `/` character or not, we can match path parts
strictly or not.

The rules for converting patterns are as follow:

- If a pattern starts with `**/` we replace this prefix by `/` and will
match the string anywhere.
- If a pattern starts with `*` we do not prefix it with `/` and will
match the string anywhere.
- If a pattern starts with `./` we replace this prefix by the directory
being watched by the current watcher and will match the string from
root.
- If a pattern starts with `/` we will match the string from root.
- If a pattern starts with none of the above, we prefix with `/` and
match the string anywhere.

Then we apply trailing conversion logic:

- If a pattern contains a `*` we remove this character plus everything
that follows.
- If a pattern ends with `/` we ensure there's only one trailing `/`.
- If a pattern does not finish by either `*` or `/` we append `/`.

The final filters are triaged between two vectors: `mRootFilters` and
`mFilters` in order to match starting from the root or to match
anywhere, respectively.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

### CODE REVIEW

**Comment by paul-marechal** [comment]
cc @vince-fugnitto @marcdumais-work

File: src/Queue.cpp [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #include "../includes/Queue.h"
+#include <cstdio>
```
Comment (paul-marechal) [inline src/Queue.cpp]: Need to remove this, not needed.

File: includes/PathFilter.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        std::vector<std::string> mFilters;
+
+    public:
+        PathFilter(const std::string &root);
+        ~PathFilter();
+
+        /** register a glob filter */
+        void addIgnoreGlob(const std::string &glob);
+        /** return true if path should be ignored, false otherwise */
+        bool ignorePath(const std::string &path);
```
Comment (vince-fugnitto) [inline includes/PathFilter.h]:
```suggestion
        bool isIgnored(const std::string &path);
```

I believe the following method can be renamed so it is clearer and more aligned with the `return` type (`bool`).

File: src/PathFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#include "../includes/PathFilter.h"
+
+#pragma unmanaged
+
+static const std::string doubleStarPrefix("**/");
+
+static bool prefix(const std::string &prefix, const std::string &str) {
```
Comment (vince-fugnitto) [inline src/PathFilter.cpp]:
```suggestion
static bool startsWith(const std::string &prefix, const std::string &str) {
```

I believe `startsWith` would be clearer.

File: src/PathFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      int i = length - 2;
+      while (i >= 0 && path[--i] == '/');
+      return path.substr(0, i + 2);
+    }
+  }
+  return path + '/';
+}
+
+static void push_unique(std::vector<std::string> &vector, const std::string &element) {
+  for (auto e : vector) {
```
Comment (vince-fugnitto) [inline src/PathFilter.cpp]:
```suggestion
static void pushUnique(std::vector<std::string> &vector, const std::string &element) {
  for (auto e : vector) {
```

Is there a standard of when to use `aB` versus `a_b` as a method name format?

**Review by vince-fugnitto — APPROVED** [review]
The changes look good to me :+1: 
I have some minor comments at your discretion.

**Comment by paul-marechal** [comment]
PR opened upstream.

**Pull request closed without merging**